src/models/partial_tx.ts

```diff
@@ -0,0 +1,440 @@
+ */
+export class PartialTx {
+  inputs: ProposalInput[];
+  outputs: ProposalOutput[];
+  network: Network;
+
+  constructor(network?: Network) {
+    this.inputs = [];
+    this.outputs = [];
+    this.network = network || new Network('mainnet');
+  }
+
+  /**
+   * Convert the PartialTx into a complete TxData ready to be signed or serialized.
+   *
+   * @throws {UnsupportedScriptError} All output scripts must be P2SH or P2PKH
+   *
+   * @return {TxData}
+   * @memberof PartialTx
+   * @inner
+   */
+  getTxData() {
+    const tokenSet = new Set<string>();
+    for (const output of this.outputs) {
+      tokenSet.add(output.token);
+    }
+    for (const input of this.inputs) {
+      tokenSet.add(input.token);
+    }
+
+    // Remove HTR from tokens array
+    tokenSet.delete(HATHOR_TOKEN_CONFIG.uid);
+    const tokens = Array.from(tokenSet);
+
+    const data = {
+      inputs: this.inputs.map(i => i.toData()),
+      outputs: this.outputs.map(o => o.toData(tokens.indexOf(o.token)+1, this.network)),
+      tokens
+    };
+
+    transaction.completeTx(data);
+
+    return data;
+  }
+
+  /**
+   * Create a Transaction instance from the PartialTx.
+   *
+   * @throws {UnsupportedScriptError} All output scripts must be P2SH or P2PKH
+   *
+   * @return {Transaction}
+   * @memberof PartialTx
+   * @inner
+   */
+  getTx() {
+    return helpers.createTxFromData(this.getTxData(), this.network);
+  }
+
+  /**
+   * Calculate balance for all tokens from inputs and outputs.
+   *
+   * @return {Record<string, number>}
+   * @memberof PartialTx
+   * @inner
+   */
+  calculateTokenBalance() {
+    const tokenBalance: Record<string, Record<string, number>> = {};
+    for (const input of this.inputs) {
+      if (!(input.token in tokenBalance)) {
+        tokenBalance[input.token] = {inputs: 0, outputs: 0};
+      }
+      tokenBalance[input.token].inputs += input.value;
+    }
+
+    for (const output of this.outputs) {
+      if (!(output.token in tokenBalance)) {
+        tokenBalance[output.token] = {inputs: 0, outputs: 0};
+      }
+
+      tokenBalance[output.token].outputs += output.value;
+    }
+
+    return tokenBalance;
+  }
+
+  /**
+   * Return true if the balance of the outputs match the balance of the inputs for all tokens.
+   *
+   * @return {Boolean}
+   * @memberof PartialTx
+   * @inner
+   */
+  isComplete() {
+    const tokenBalance = this.calculateTokenBalance();
+
+    // Calculated the final balance for all tokens
+    // return if all are 0
+    return Object.values(tokenBalance).every(v => v.inputs === v.outputs);
+  }
+
+  /**
+   * Add an UTXO as input on the PartialTx.
+   * This method is async because we need to fetch information on the inputs (token, value)
+   *
+   * @param {string} txId The transaction id of the UTXO.
+   * @param {number} index The index of the UTXO.
+   *
+   * @return {Promise<void>}
+   * @memberof PartialTx
+   * @inner
+   */
+  async addInput(txId: string, index: number) {
```

Comment:
An user could add an input by the `tx_id` and `index` which is not from his wallet.
Depending on how the negotiation process is implemented this could be a likely scenario.
I thought of searching the local history before calling the `txApi` but it would only be an early optimization and not change the overall feature.

Reply:
How can he add a utxo that is not from his wallet? How is he going to sign those inputs?

Reply:
He doesn't sign it, he can add it and send to the other wallet to sign.

I think this rare case was a bad example though, we have to remember that a serialized partial tx is passed among participants and the wallet will need to calculate the balance and show inputs that the other participants add (i.e. not from the wallet).

Reply:
As discussed previously, the call signatures changed so that adding inputs always needs the metadata information (value, token, authorities) and this metadata is added to the `PartialTx` serialization so that it is not lost during the negotiation.

A validation method was added to the `PartialTx`  which calls the fullnode and checks that the inputs metadata are valid.
This is called before signing so we never sign an invalid transaction but the fullnode is not called during the negotiation phase anymore.